Support chat template and `echo` for chat API

Tostino opened the pull request:
This pull request introduces the chat template feature to vLLM, utilizing the template stored in the tokenizer, enhancing its compatibility with the OpenAI Chat API.
https://huggingface.co/blog/chat-templates

This only affects the OpenAI API `chat/completion` endpoint, the regular `completion` endpoint does not utilize this feature.

There has already been a ton of discussion under this PR: https://github.com/vllm-project/vllm/pull/1493 but I accidentally messed things up by replacing the branch, so we are trying this again...

1. Addition of the `--chat-template` command-line argument to specify a chat template file or single-line template for the model.
2. Implementation of the `--response-role` command-line argument for defining the role name in chat responses when `add_generation_prompt` is set to true.
3. Update to the chat API request handling to support handling finishing a partial response correctly, and echoing input portions of messages (request.add_generation_prompt, and request.echo).
4. Addition of new chat templates examples (`template_chatml.jinja`, `template_alpaca.jinja`, and `template_inkbot.jinja`) showing the multiple ways they can be specified.
5. More robust error handling, and fix the responses to actually match the OpenAI API format.
6. Update quickstart.rst to show the new features.

Tostino committed:
Enhance vLLM server with chat template support.

1. Addition of the `--chat-template` command-line argument to specify a chat template file or single-line template for the model.
2. Implementation of the `--response-role` command-line argument for defining the role name in chat responses when `add_generation_prompt` is set to true.
3. Introduction of the `--disable-endpoints` argument to allow disabling of specific server endpoints.
4. Update to the chat API request handling to support handling finishing a partial response correctly, and echoing input portions of messages (request.add_generation_prompt, and request.echo).
5. Addition of new chat templates in JSON and Jinja formats (`template_chatml.json`, `template_alpaca.jinja`, and `template_inkbot.jinja`) showing the multiple ways they can be specified.
6. More robust error handling, and fix the responses to actually match the OpenAI API format.
7. Update quickstart.rst to show the new features.

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -640,6 +691,10 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
     else:
         served_model = args.model
 
+    disabled_endpoints = args.disable_endpoints
```
This seems like a separate features no?

Tostino replied:
Yes at this point since I couldn't implement a hard-fail on loading templates at startup easily. I mentioned it in the other PR last night that I left it in place because I had already implemented it at Simon's request. I can tear that out if necessary.

Tostino replied:
Removed.

Tostino commented on examples/template_chatml.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Okay, I'll convert this one over to a jinja file.

Tostino replied:
Done. Simplified the loading code as well.

aarnphm commented:
@simon-mo whats your opinion on just supporting chat templates through envvar?

vllm should be able to document how people can pass in the jinja templates through a envvar, so that vllm won't handle any parsing, and users are responsible for this?

I think this is powerful in a way that users have full control of the chat templates. By default, if there is none provided then fall back to huggingface behaviour?

I think we can provide a default chat templates if needed

Tostino committed:
Remove support for --disable-endpoints.

Tostino committed:
Remove template_chatml.json and replace with template_chatml.jinja, and simplify the template loading code to remove support for json based templates.

Tostino commented:
> Can we also make sure there is a way to disable chat templates such that users provide messages that are preformatted?

As said before, you use the regular `completion` endpoint if you want to provide a pre-formatted string to the model.

Explain how we would provide the preformatted text using the chat/completion endpoint which requires a list of messages.
Here is an example call:
```
curl http://0.0.0.0:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer " \
  -d '{
    "model": "Tostino/Inkbot-13B-8k-0.2",
    "stop": ["\n<#user#>", "\n<#bot#>"],
    "stream": false,
    "add_generation_prompt": true,
    "echo": true,
    "n": 1,
    "messages": [
	{"role": "meta-current_date", "content": "2023-10-20"},
	{"role": "meta-task_name", "content": "kg"},
	{"role": "system", "content": "You are a helpful assistant."},
	{"role": "user", "content": "Hello!"},
	{"role": "assistant", "content": "Hello, how are you?"},
	{"role": "user", "content": "Great, thank you! Now i would like some help planning my weekend in Asheville, I would like some hike suggestions."},
	{"role": "assistant", "content": "Asheville is a fantastic destination for hiking enthusiasts! Here are a few suggestions:\n\n1. Black Balsam Knob Trail: This 4.4-mile round trip trail offers stunning views of the surrounding mountains.\n2. Max Patch Loop: This 2.4-mile loop is perfect for families or beginners, featuring wildflowers and mountain views.\n3. Shut-ins Trail: The 2.8-mile hike takes you through a scenic gorge with cascading waterfalls.\n4. Graveyard Fields: For a longer hike, this 8-mile loop takes you through meadows, streams, and a beautiful forest.\n5. Linville Gorge: If you are up for a more challenging hike, the Linville Gorge has various trails with breathtaking views."},
	{"role": "user", "content": "Extract key concepts and relationships from the conversation to form a knowledge graph."}	
	]
  }'
  ```

casper-hansen commented:
I see where you are coming from - perhaps it should be mentioned in the PR that this only applies to certain endpoints.

aarnphm commented:
> > Can we also make sure there is a way to disable chat templates such that users provide messages that are preformatted?
> 
> As said before, you use the regular `completion` endpoint if you want to provide a pre-formatted string to the model.
> 
> Explain how we would provide the preformatted text using the chat/completion endpoint which requires a list of messages. Here is an example call:
> 
> ```
> curl http://0.0.0.0:8000/v1/chat/completions \
>   -H "Content-Type: application/json" \
>   -H "Authorization: Bearer " \
>   -d '{
>     "model": "Tostino/Inkbot-13B-8k-0.2",
>     "stop": ["\n<#user#>", "\n<#bot#>"],
>     "stream": false,
>     "add_generation_prompt": true,
>     "echo": true,
>     "n": 1,
>     "messages": [
> 	{"role": "meta-current_date", "content": "2023-10-20"},
> 	{"role": "meta-task_name", "content": "kg"},
> 	{"role": "system", "content": "You are a helpful assistant."},
> 	{"role": "user", "content": "Hello!"},
> 	{"role": "assistant", "content": "Hello, how are you?"},
> 	{"role": "user", "content": "Great, thank you! Now i would like some help planning my weekend in Asheville, I would like some hike suggestions."},
> 	{"role": "assistant", "content": "Asheville is a fantastic destination for hiking enthusiasts! Here are a few suggestions:\n\n1. Black Balsam Knob Trail: This 4.4-mile round trip trail offers stunning views of the surrounding mountains.\n2. Max Patch Loop: This 2.4-mile loop is perfect for families or beginners, featuring wildflowers and mountain views.\n3. Shut-ins Trail: The 2.8-mile hike takes you through a scenic gorge with cascading waterfalls.\n4. Graveyard Fields: For a longer hike, this 8-mile loop takes you through meadows, streams, and a beautiful forest.\n5. Linville Gorge: If you are up for a more challenging hike, the Linville Gorge has various trails with breathtaking views."},
> 	{"role": "user", "content": "Extract key concepts and relationships from the conversation to form a knowledge graph."}	
> 	]
>   }'
> ```

I think chat templates should only apply for `/v1/chat/completions`. Users can change the chat templates if needed. `/v1/completions` should still stay the way they were before, which means that users should have full responsibility for formatting their prompt.

intrafindBreno commented:
This pull request is a great contribution to the vllm project. I hope it gets merged soon!

dongxiaolong commented:
Hi @Tostino ,

Thank you so much for your PR! Your contribution has successfully addressed key issues in model inference, enabling the implementation of function calls using models like OpenHermes. Here is an example I created that demonstrates the effective application of OpenHermes: [OpenHermes Functions with VLLM](https://github.com/dongxiaolong/open-source-function-calling/blob/main/openhermes-functions-with-vllm.ipynb). Your work has truly unlocked the potential of OpenHermes. Adding another practical example to your contribution would greatly help others understand the significant impact of this work.

tjtanaa commented:
Huggingface now supports `chat_template` through their tokenizer class. I think if the community are using this feature, it will make chat template tracking easier. 

To that end, we have setup a [huggingface space](https://huggingface.co/spaces/EmbeddedLLM/chat-template-generation) that allows developers and the community to increase the coverage of models to make use of this feature. We have provided many of the well-known prompt templates. There are many more templates to come within these few weeks.

The ui interface also allows user to download chat template as jinja2 which I think would be beneficial for this PR feature as well.

Tostino commented:
@tjtanaa That is a neat tool, glad there is more work going into this in the same direction. Being able to download the template to a file easily would be beneficial for users.
One thing to note, is that it incorrectly parses my template for Tostino/Inkbot-13B-8k-0.2 when I tried. The newlines are ignored.
![image](https://github.com/vllm-project/vllm/assets/8227578/1084be37-aae0-4f1f-a367-305e199d637e)


@dongxiaolong Glad you found it useful! Very cool example to see working.

Tostino commented:
@tjtanaa where can we discuss that issue other than on this PR?

My template is valid jinja, and it does have new lines embedded in the single-line version of it. There was an extra step to correctly load the single line jinja in this PR to deal with new lines properly, I'm guessing you just need to do the same.

Tostino commented:
@aarnphm / @simon-mo  / @WoosukKwon (or anyone else appropriate)

Is there anything that needs to be addressed prior to merging this PR at this point? There is nothing else I am aware of for this PR, as it's been pretty thoroughly tested at this point.

aarnphm commented on docs/source/getting_started/quickstart.rst:
```diff
@@ -122,14 +123,22 @@ Use model from www.modelscope.cn
 
     $ curl http://localhost:8000/v1/models
 
+Using OpenAI Completions API with vLLM
+-------------------------------
```
```suggestion
Using OpenAI Completions API with vLLM
--------------------------------------
```

aarnphm commented on examples/template_chatml.jinja:
```diff
@@ -0,0 +1,3 @@
+{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}
```
I don't think this logic is correct here. since `add_generation_prompt` is always available within the jinja environment for `apply_chat_template`

Tostino replied:
That was a copy/paste from...I can't recall where, but I have absolutely seen quite a few templates do the same. A google for that string shows how widely it's spread.

Agreed that it seems redundant.

Tostino replied:
Alright, removed.

Tostino committed:
Update docs/source/getting_started/quickstart.rst

Co-authored-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -669,6 +670,17 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
+                        help="The file path to the chat template, "
+                        "or the template in single-line form "
+                        "for the specified model")
+    parser.add_argument("--response-role",
```
I think this largely depends on the templates itself right? By default I don't think this is needed.

Tostino replied:
The response-role? Some models may not use user/assistant as the role names. The templates are free with that. I defaulted to the same behavior as the OpenAI API.

aarnphm replied:
Ye but we already have that in the message['role'] right? not sure why we need this

Tostino replied:
Yes it usually alternates between user/assistant...but given the following request, how would we know the role to respond with?
```
curl http://0.0.0.0:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer " \
  -d '{
    "model": "Tostino/Inkbot-13B-8k-0.2",
    "stream": false,
    "add_generation_prompt": true,
    "echo": true,
    "n": 1,
    "messages": [
	{"role": "meta-current_date", "content": "2023-10-20"},
	{"role": "meta-task_name", "content": "general"},
	{"role": "system", "content": "You are a helpful assistant."},
	{"role": "user", "content": "What is the capital of the USA?"}	
	]
  }'
  ```
  
Previously, it was hard-coded as `assistant`. Now it is configurable with that cli argument. Ideally, this would be additional metadata about how to use the template that would be stored somewhere in the tokenizer...but HF didn't think that far ahead.

aarnphm replied:
Users should be responsible for doing few-shot prompt right?

Tostino replied:
No, that is totally unnecessary for a lot of models and just eats up context space.

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -250,128 +223,155 @@ async def create_chat_completion(request: ChatCompletionRequest,
+        return response
 
+    # Streaming response
     if request.stream:
```
actually, I'm always curious about this logic. request.stream is True, then this would never become reachable right?

Tostino replied:
With the old logic, it was dead code. With the new logic with them both broken out into their own functions, it's not dead code any more.

Tostino committed:
Update template_chatml.jinja

Remove redundant definition of add_generation_prompt from template. This was in one of the templates I found for chatml, but it is unnecessary.

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -250,128 +223,155 @@ async def create_chat_completion(request: ChatCompletionRequest,
-
-        return response_json
+    def get_role() -> str:
+        if request.add_generation_prompt:
+            return response_role
+        else:
+            return request.messages[-1]["role"]
```
Like here should it just be `request.messages[-1]['role']`?

Tostino replied:
No, that doesn't work.

Yard1 commented:
@Tostino Would it be possible to add some simple unit tests for this?

Tostino commented:
Sure, are there any existing tests for the server that I can add to?

Yard1 commented:
I don't think there are any for the OpenAI server specifically.

Tostino commented:
@Yard1 So, to properly test this, it looks like I need to refactor a whole lot more code...are you sure you want me to do that? I am not doing any more work that will be thrown away...I've wasted far too much time on this already.

And it now looks like i'll have to spend more time rebasing, because there are conflicts again.

Tostino commented:
Well, time to open a new PR... The merge didn't go cleanly, and there is no way to force-push without closing this that I know of.

Tostino committed:
Fix merge issue

aarnphm commented:
@Tostino probably best to add integration tests for the openapi server.

simon-mo commented:
This PR is already pretty sizable. I'm leaning towards merging it (as part of release for v0.2.3 and have testing as immediate follow up.

Yard1 commented:
Hmm merging without tests doesn't seem right to me but if you think that's fine I won't block on it :)

simon-mo commented:
Thank you for adding the tests.

I made a pass for small code quality nits. Attached diff in this comment. @Tostino please let me know when you have finished adding tests so I can directly push to this branch without causing you more merge conflict and resolving them one by one. 

I think for the future work, the design considerations still need to flush out with actual use cases. I'm aware of previous discussion on this, this PR should not be blocked by any of these. Just putting these here as notes:
* `add_generation_prompt`: is it really a request parameter or bound by the model itself. I have the feeling that most use case will always set it to one value and doesn't change across requests.
* `echo` for chat completion API: currently this PR prepend the previous role is the same (essentially a "continue" type of workflow during chat); because openai doesn't have the `echo` semantics defined for chat, we can think a bit more about exactly what's the semantics; and be aware of future changes. 

<details><summary>Diff Attached</summary>
<p>

```diff
diff --git a/docs/source/getting_started/quickstart.rst b/docs/source/getting_started/quickstart.rst
index 1516a7b..bd3940f 100644
--- a/docs/source/getting_started/quickstart.rst
+++ b/docs/source/getting_started/quickstart.rst
@@ -124,11 +124,12 @@ Use model from www.modelscope.cn
     $     --model="qwen/Qwen-7B-Chat" --revision="v1.1.8" --trust-remote-code
 
 By default, the server uses a predefined chat template stored in the tokenizer. You can override this template by using the ``--chat-template`` argument:
+
 .. code-block:: console
 
    $ python -m vllm.entrypoints.openai.api_server \
-       --model facebook/opt-125m \
-       --chat-template ./examples/template_chatml.json
+   $     --model facebook/opt-125m \
+   $     --chat-template ./examples/template_chatml.json
 
 This server can be queried in the same format as OpenAI API. For example, list the models:
 
@@ -137,7 +138,7 @@ This server can be queried in the same format as OpenAI API. For example, list t
     $ curl http://localhost:8000/v1/models
 
 Using OpenAI Completions API with vLLM
---------------------------------------
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Query the model with input prompts:
 
@@ -167,7 +168,7 @@ Since this server is compatible with OpenAI API, you can use it as a drop-in rep
 For a more detailed client example, refer to `examples/openai_completion_client.py <https://github.com/vllm-project/vllm/blob/main/examples/openai_completion_client.py>`_.
 
 Using OpenAI Chat API with vLLM
--------------------------------
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 The vLLM server is designed to support the OpenAI Chat API, allowing you to engage in dynamic conversations with the model. The chat interface is a more interactive way to communicate with the model, allowing back-and-forth exchanges that can be stored in the chat history. This is useful for tasks that require context or more detailed explanations.
 
diff --git a/vllm/entrypoints/openai/api_server.py b/vllm/entrypoints/openai/api_server.py
index b87fcd9..54ff3c8 100644
--- a/vllm/entrypoints/openai/api_server.py
+++ b/vllm/entrypoints/openai/api_server.py
@@ -47,15 +47,22 @@ def create_error_response(status_code: HTTPStatus,
                         status_code=status_code.value)
 
 
-def load_chat_template():
-    try:
-        with open(args.chat_template, "r") as f:
-            content = f.read()
-            return content
-    except OSError:
-        # If opening a file fails, set chat template to be args to
-        # ensure we decode so our escape are interpreted correctly
-        return codecs.decode(args.chat_template, "unicode_escape")
+def load_chat_template(args, tokenizer):
+    if args.chat_template is not None:
+        try:
+            with open(args.chat_template, "r") as f:
+                chat_template = f.read()
+        except OSError:
+            # If opening a file fails, set chat template to be args to
+            # ensure we decode so our escape are interpreted correctly
+            chat_template = codecs.decode(args.chat_template, "unicode_escape")
+
+        tokenizer.chat_template = chat_template
+        logger.info(f"Using supplied chat template:\n{tokenizer.chat_template}")
+    elif tokenizer.chat_template is not None:
+        logger.info(f"Using default chat template:\n{tokenizer.chat_template}")
+    else:
+        logger.warning("No chat template provided. Chat API will not work.")
 
 
 @app.exception_handler(RequestValidationError)
@@ -73,16 +80,6 @@ async def check_model(request) -> Optional[JSONResponse]:
     return ret
 
 
-async def get_gen_prompt(request) -> str:
-    try:
-        return tokenizer.apply_chat_template(
-            conversation=request.messages,
-            tokenize=False,
-            add_generation_prompt=request.add_generation_prompt)
-    except Exception as e:
-        raise RuntimeError(f"Error generating prompt: {str(e)}") from e
-
-
 async def check_length(
     request: Union[ChatCompletionRequest, CompletionRequest],
     prompt: Optional[str] = None,
@@ -174,8 +171,6 @@ async def create_chat_completion(request: ChatCompletionRequest,
         - function_call (Users should implement this by themselves)
         - logit_bias (to be supported by vLLM engine)
     """
-    logger.info(f"Received chat completion request: {request}")
-
     error_check_ret = await check_model(request)
     if error_check_ret is not None:
         return error_check_ret
@@ -186,9 +181,12 @@ async def create_chat_completion(request: ChatCompletionRequest,
                                      "logit_bias is not currently supported")
 
     try:
-        prompt = await get_gen_prompt(request)
-    except RuntimeError as e:
-        logger.error(f"Error in generating prompt from request: {str(e)}")
+        prompt = tokenizer.apply_chat_template(
+            conversation=request.messages,
+            tokenize=False,
+            add_generation_prompt=request.add_generation_prompt)
+    except Exception as e:
+        logger.error(f"Error in applying chat template from request: {str(e)}")
         return create_error_response(HTTPStatus.BAD_REQUEST, str(e))
 
     token_ids, error_check_ret = await check_length(request, prompt=prompt)
@@ -198,7 +196,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
     model_name = request.model
     request_id = f"cmpl-{random_uuid()}"
     created_time = int(time.monotonic())
-    obj_str = "chat.completion.chunk"
+    chunk_object_type = "chat.completion.chunk"
     try:
         spaces_between_special_tokens = request.spaces_between_special_tokens
         sampling_params = SamplingParams(
@@ -236,7 +234,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
             choice_data = ChatCompletionResponseStreamChoice(
                 index=i, delta=DeltaMessage(role=role), finish_reason=None)
             chunk = ChatCompletionStreamResponse(id=request_id,
-                                                 object=obj_str,
+                                                 object=chunk_object_type,
                                                  created=created_time,
                                                  choices=[choice_data],
                                                  model=model_name)
@@ -258,7 +256,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
                         delta=DeltaMessage(content=last_msg_content),
                         finish_reason=None)
                     chunk = ChatCompletionStreamResponse(id=request_id,
-                                                         object=obj_str,
+                                                         object=chunk_object_type,
                                                          created=created_time,
                                                          choices=[choice_data],
                                                          model=model_name)
@@ -273,8 +271,12 @@ async def create_chat_completion(request: ChatCompletionRequest,
             res: RequestOutput
             for output in res.outputs:
                 i = output.index
-                # Send token-by-token response for each request.n
-                if output.finish_reason is None and not finish_reason_sent[i]:
+
+                if finish_reason_sent[i]:
+                    continue
+
+                if output.finish_reason is None:
+                    # Send token-by-token response for each request.n
                     delta_text = output.text[len(previous_texts[i]):]
                     previous_texts[i] = output.text
                     completion_tokens = len(output.token_ids)
@@ -284,15 +286,14 @@ async def create_chat_completion(request: ChatCompletionRequest,
                         delta=DeltaMessage(content=delta_text),
                         finish_reason=None)
                     chunk = ChatCompletionStreamResponse(id=request_id,
-                                                         object=obj_str,
+                                                         object=chunk_object_type,
                                                          created=created_time,
                                                          choices=[choice_data],
                                                          model=model_name)
                     data = chunk.json(exclude_unset=True, ensure_ascii=False)
                     yield f"data: {data}\n\n"
-                # Send the finish response for each request.n only once
-                if output.finish_reason is not None and not finish_reason_sent[
-                        i]:
+                else:
+                    # Send the finish response for each request.n only once
                     prompt_tokens = len(res.prompt_token_ids)
                     final_usage = UsageInfo(
                         prompt_tokens=prompt_tokens,
@@ -302,7 +303,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
                     choice_data = ChatCompletionResponseStreamChoice(
                         index=i, delta=[], finish_reason=output.finish_reason)
                     chunk = ChatCompletionStreamResponse(id=request_id,
-                                                         object=obj_str,
+                                                         object=chunk_object_type,
                                                          created=created_time,
                                                          choices=[choice_data],
                                                          model=model_name)
@@ -326,6 +327,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
                                              "Client disconnected")
             final_res = res
         assert final_res is not None
+
         choices = []
         role = get_role()
         for output in final_res.outputs:
@@ -710,14 +712,8 @@ if __name__ == "__main__":
         engine_model_config.tokenizer,
         tokenizer_mode=engine_model_config.tokenizer_mode,
         trust_remote_code=engine_model_config.trust_remote_code)
+    load_chat_template(args, tokenizer)
 
-    chat_template = None
-    if args.chat_template is not None:
-        chat_template = load_chat_template()
-    if chat_template is not None:
-        tokenizer.chat_template = chat_template
-    if tokenizer.chat_template is not None:
-        logger.info(f"Using chat template:\n{tokenizer.chat_template}")
 
     uvicorn.run(app,
                 host=args.host,
```

</p>
</details>

Tostino committed:
Add a few tests, and refactor the methods required to support that.
Fixed issues with chatml template not actually supporting the add_generation_prompt feature. This was just a copy/paste from a random model.

Tostino commented:
Thank you very much @simon-mo. I just added a handful of tests, and fixed the chatml template after tests identified some issues.  Should be ready for you now.

simon-mo committed:
readability pass

simon-mo committed:
fix tests

simon-mo committed:
lint

simon-mo commented:
Oh one more future work could be loading chat template from http url. Let's see if that will be a common request and decide whether to be added.

flexchar commented:
Hey all, I've been following the new and old thread for weeks, if not months, now. I wanted to tell that this is greatly awaited and appreciated work. Thank you @simon-mo and @Tostino for getting this into the framework.

Do we have an expected date to pin the new version for vLLM? :)

Tostino commented:
@flexchar https://github.com/vllm-project/vllm/issues/1856

Should be pretty soon.

flexchar commented:
I was waiting for this PR so long but I found out that making request to `/chat/completion` vs `/completion` where I manually prepare prepare ChatML string, ends up costing 2x as long. My input is 1000 tokens and output 100 tokens. @Tostino any ideas why that could be?

Tostino commented:
> I was waiting for this PR so long but I found out that making request to `/chat/completion` vs `/completion` where I manually prepare prepare ChatML string, ends up costing 2x as long. My input is 1000 tokens and output 100 tokens. @Tostino any ideas why that could be?

@flexchar Not off the top of my head, i've not had any noticeable slowdown between `chat/completions` and `completions`. I do notice a slowdown when using streaming (~10-20%), but nothing like a 2x slowdown.

I'm getting ~300-500 tokens/sec generated throughput using 1x 3090 and a llama2 13b model using the `chat/completions` endpoint.

Do you have an example you can share to trigger it?

flexchar commented:
Thank you for trying. I don't have an easy reproducable example but the next time I work on that part I will do one certainly. I appreciate you testing thou :)

PeterXiaTian commented:
vllm/entrypoints/api_server.py　　　这里为什么还没有这个参数了？

PhilipMay commented:
How do I use the chat template with the "offline inference" with `vllm.LLM`?

This PR only enables this for the REST API as far as I can see. @Tostino

Tostino commented:
Sorry, on mobile right now so going from memory. I believe that there wasn't a "local" equivalent of the `chat/completions` API when I implemented this. So it was implemented for the REST endpoint, because that's all it could work with unless I did a bunch of extra work to also add an equivalent local version of the chat completions API.

PeterXiaTian commented:
well，i see，thanks发自我的 iPhone在 2024年4月25日，23:11，Adam Brusselback ***@***.***> 写道：﻿
Sorry, on mobile right now so going from memory. I believe that there wasn't a "local" equivalent of the chat/completions API when I implemented this. So it was implemented for the REST endpoint, because that's all it could work with unless I did a bunch of extra work to also add an equivalent local version of the chat completions API.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>
